packages/@aws-cdk/core/lib/stack.ts

```diff
@@ -734,32 +736,113 @@ export class Stack extends Construct implements ITaggable {
-      throw new Error(`'${target.node.path}' depends on '${this.node.path}' (${cycle.join(', ')}). Adding this dependency (${reason}) would create a cyclic reference.`);
+      throw new Error(`'${target.node.path}' depends on '${this.node.path}' (${cycle.map(x=>x.description).join(', ')}). Adding this dependency (${reason.description}) would create a cyclic reference.`);
     }
 
     let dep = this._stackDependencies[Names.uniqueId(target)];
     if (!dep) {
-      dep = this._stackDependencies[Names.uniqueId(target)] = {
-        stack: target,
-        reasons: [],
-      };
+      dep = this._stackDependencies[Names.uniqueId(target)] = { stack: target, reasons: [] };
+    }
+    let reasons = filterReasons(dep, reason);
+    if (reasons.length > 0) {
+      // Dependency already exists and for the provided reason
+      return;
     }
-
     dep.reasons.push(reason);
 
     if (process.env.CDK_DEBUG_DEPS) {
       // eslint-disable-next-line no-console
-      console.error(`[CDK_DEBUG_DEPS] stack "${this.node.path}" depends on "${target.node.path}" because: ${reason}`);
+      console.error(`[CDK_DEBUG_DEPS] stack "${reason.source.node.path}" depends on "${reason.target.node.path}" because: ${reason.description}`);
+    }
+  }
+
+  /**
+   * Called implicitly by the `obtainDependencies` helper function in order to
+   * collect resource dependencies across two top-level stacks at the assembly level.
+   *
+   * Use `stack.obtainDependencies` to see the dependencies between any two stacks.
+   *
+   * @internal
+   */
+  public _obtainAssemblyDependencies(reasonFilter: StackDependencyReason={ description: '' }) {
+    let dependencies: Set<Element> = new Set();
+    Object.values(this._stackDependencies).forEach((dep) => {
+      filterReasons(dep, reasonFilter).forEach(r => {
+        if (r.target) {
+          dependencies.add(r.target);
+        }
+      });
+    });
+    return Array.from(dependencies);
+  }
+
+  /**
+   * Called implicitly by the `removeDependency` helper function in order to
+   * remove a dependency between two top-level stacks at the assembly level.
+   *
+   * Use `stack.addDependency` to define the dependency between any two stacks,
+   * and take into account nested stack relationships.
+   *
+   * @internal
+   */
+  public _removeAssemblyDependency(target: Stack, reasonFilter: StackDependencyReason={ description: 'dependency removed using stack.removeDependency()' }) {
+    // defensive: we should never get here for nested stacks
+    if (this.nested || target.nested) {
+      throw new Error('There cannot be assembly-level dependencies for nested stacks');
+    }
+    // No need to check for a dependency cycle when removing one
+
+    // Fill in reason details if not provided
+    if (!reasonFilter.source) {
+      reasonFilter.source = this;
+    }
+    if (!reasonFilter.target) {
+      reasonFilter.target = target;
+    }
+
+    let dep = this._stackDependencies[Names.uniqueResourceName(target, {})];
```

Comment:
While @comcalvi is right that we're moving in this direction, I'm not sure it will work in this context. Because we're still using `uniqueId` throughout the rest of this, if you're looking for their unique resource name instead here, it may come up empty. I I think this exemplifies the need for integration tests on this change.

Reply:
Oh, also, you can totally just do `Names.uniqueResourceName(target)` if/when using this and omit the `{}` altogether. It's optional.

Reply:
Ok, I'm working on some better tests, including integration as well as thinking through the `uniqueResourceName` vs `uniqueId` issue.

It looks like I'll have to follow the `@aws-cdk/core/test/integration.custom-resources.readme` pattern and actually place it under `@aws-cdk/aws-cloudformation`.

Also - TypeScript doesn't seem to agree about `options` being optional:
```lib/stack.ts:819:45 - error TS2554: Expected 2 arguments, but got 1.

819     let dep = this._stackDependencies[Names.uniqueResourceName(target)];
                                                ~~~~~~~~~~~~~~~~~~~~~~~~~~

  lib/names.ts:84:59
    84   public static uniqueResourceName(construct: IConstruct, options: UniqueResourceNameOptions) {
                                                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    An argument for 'options' was not provided.
```

Reply:
Lol, I wrote that code and don't properly remember what I actually did there. You'e right on this.